presto-main/src/main/java/com/facebook/presto/execution/scheduler/nodeSelection/SimpleTtlNodeSelector.java

```diff
@@ -162,14 +162,9 @@ public List<InternalNode> selectRandomNodes(int limit, Set<InternalNode> exclude
         List<InternalNode> eligibleNodes = filterNodesByTtl(activeNodes, excludedNodes, ttlInfo, estimatedExecutionTimeRemaining);
         List<InternalNode> selectedNodes = selectNodes(limit, new ResettableRandomizedIterator<>(eligibleNodes));
 
-        if (selectedNodes.isEmpty()) {
-            if (fallbackToSimpleNodeSelection) {
-                log.warn("No nodes available with enough TTL (%s), falling back to simple node selection.", estimatedExecutionTimeRemaining);
-                return simpleNodeSelector.selectRandomNodes(limit, excludedNodes);
-            }
-
-            log.warn("No nodes available with enough TTL (%s). Active nodes: %s", estimatedExecutionTimeRemaining, activeNodes);
-            throw new PrestoException(NO_NODES_AVAILABLE, "No nodes available to run query");
+        if (selectedNodes.isEmpty() && fallbackToSimpleNodeSelection) {
+            log.warn("No nodes available with enough TTL (%s), falling back to simple node selection.", estimatedExecutionTimeRemaining);
+            return simpleNodeSelector.selectRandomNodes(limit, excludedNodes);
         }
 
         return selectedNodes;
```

Comment:
So now if fallbackToSimpleNodeSelection is false, we will allow returning an empty set of nodes back to the caller ? Why was that not allowed earlier ? How would the caller deal with that ?

Reply:
If we look at the `SimpleNodeSelector`, on which the `SimpleTtlNodeSelector` is based, the `selectRandomNodes` function can indeed return an empty list and the higher level function utilizing this has the context on whether an empty list is okay or not. I think the behavior should be consistent with what is expected from `selectRandomNodes`. It shouldn’t make decisions when it doesn’t have the full context. There are two cases when the `selectRandomNodes` nodes in TTL scheduling can return empty:

1. `excludedNodes` includes all workers. In that case, whether we fallback or not, we still return an empty list.
2. Good-TTL nodes are not available at all. In which case, if we fallback, we can return some nodes. If fallback is not enabled, we return empty, and the higher level caller will raise an error if needed.